# Libro de análisis · Empleos en CV (mini investigación)

**Datasets de curriculums procesados → ¿qué insight sobre la vida laboral?**

- Este libro responde preguntas de negocio/analíticas sobre **376.567 experiencias
  laborales** (de 71.061 personas) del archivo limpio
  `data/03_processed/empleos_limpio.csv`.
- Es **solo lectura**: ninguna celda modifica el CSV ni guarda derivados.
  Toda métrica se **calcula en vivo** y ninguna cifra se escribe a mano.
- Reproduce todo de cero con *Kernel → Restart & Run All*. Las cifras de la
  línea base del bloque 1 se validan contra el `README.md`, que es la
  referencia versionada del proyecto. La auditoría de la rama (`docs/AUDITORIA_RAMA_PRUEBA.md`) está **fuera de git** y documenta una versión anterior del pipeline (era `train`, 1.506.445 filas); sirve como contexto, no como fuente de verdad.


In [1]:
from pathlib import Path
import sys

import numpy as np
import pandas as pd

try:
    from IPython.display import display
except ImportError:  # ejecución fuera de Jupyter (validación automática)
    def display(*objetos, **kwargs):
        for o in objetos:
            if hasattr(o, "to_string"):
                print(o.to_string())
            else:
                print(o)

PROYECTO = Path.cwd()
while not (PROYECTO / "data").exists() and PROYECTO != PROYECTO.parent:
    PROYECTO = PROYECTO.parent
sys.path.insert(0, str(PROYECTO / "src" / "filtro"))
import filtros as F
import indicadores as I

RUTA = PROYECTO / "data" / "03_processed" / "empleos_limpio.csv"
# dtype=str: los códigos ESCO son texto y deben conservar sus ceros a la
# izquierda ('0110' no puede colisionar con '110'). Es la convención del
# proyecto, no un descuido: los contadores de competencias también llegan
# como texto y se convierten con pd.to_numeric donde se usan.
empleos = pd.read_csv(
    RUTA, dtype=str, encoding="utf-8", keep_default_na=False, na_values=[""]
)
print("Raíz del proyecto:", PROYECTO)
print("Archivo           :", RUTA.name, f"({empleos.shape[0]:,} filas x {empleos.shape[1]} columnas)")

Raíz del proyecto: C:\Users\ACER\Desktop\MineriaProtect
Archivo           : empleos_limpio.csv (376,567 filas x 20 columnas)


## 📦 1 · Conocer el dataset

Primero fijamos la **línea base**: si estas cifras no coinciden con la auditoría,
algo cambió en el pipeline y este libro no es interpretable. Luego describimos
periodo, niveles, ocupaciones y banderas de clasificación.


In [2]:
filas, columnas = empleos.shape
personas = int(empleos["resume_id"].nunique())
# Las banderas es_* se leen como texto ('True'/'False') por la regla dtype=str.
# F.contar_banderas normaliza; sumar o promediar la columna en crudo falla o
# devuelve NaN en silencio (ver notas de F.contar_banderas).
bandera = lambda col: F.contar_banderas(empleos, col)
linea_base = pd.DataFrame({
    "metrica": [
        "filas", "columnas", "personas", "es_unknown_ocupacion",
        "es_rescatado", "es_vigente", "university_level='No reportado'",
        "saber_skills='ok'",
    ],
    "valor": [
        filas, columnas, personas,
        bandera("es_unknown_ocupacion"),
        bandera("es_rescatado"),
        bandera("es_vigente"),
        int(empleos["university_level"].eq("No reportado").sum()),
        int(empleos["saber_skills"].eq("ok").sum()),
    ],
})
display(linea_base)

# Validación contra la auditoría: el libro solo tiene sentido si esto pasa.
assert filas == 376_567, f"Filas esperadas 376.567, se hallaron {filas}"
assert columnas == 20, f"Columnas esperadas 20, se hallaron {columnas}"
assert personas == 71_061, f"Personas esperadas 71.061, se hallaron {personas}"
assert bandera("es_unknown_ocupacion") == 25_805
assert bandera("es_rescatado") == 2_499
assert bandera("es_vigente") == 18_956
assert int(empleos["university_level"].eq("No reportado").sum()) == 42_807
assert int(empleos["saber_skills"].eq("ok").sum()) == 348_263
print("Línea base correcta: coincide con la auditoría.")

,metrica,valor
0,filas,376567
1,columnas,20
2,personas,71061
3,es_unknown_ocupacion,25805
4,es_rescatado,2499
5,es_vigente,18956
6,university_level='No reportado',42807
7,saber_skills='ok',348263


Línea base correcta: coincide con la auditoría.


In [3]:
inicios = empleos["start_date"].min(), empleos["start_date"].max()
fines = empleos.loc[empleos["end_date"].ne("Present"), "end_date"]
print("Rango de inicios              :", inicios)
print("Rango de fines (sin Present)  :", (fines.min(), fines.max()))
vigentes_present = int(empleos["end_date"].eq("Present").sum())
print("Experiencias vigentes ('Present'):", vigentes_present)


Rango de inicios              :

 ('Q1 1965', 'Q4 2020')
Rango de fines (sin Present)  : ('Q1 1970', 'Q4 2024')
Experiencias vigentes ('Present'): 18956


In [4]:
niveles = empleos["university_level"].value_counts().to_frame("experiencias")
niveles["personas"] = empleos.groupby("university_level")["resume_id"].nunique()
niveles["%_del_total"] = (niveles["experiencias"] / filas * 100).round(2)
display(niveles)


,experiencias,personas,%_del_total
university_level,,,
Secondary school,147904,28939,39.28
Bachelor,129507,22854,34.39
Master,54697,10263,14.53
No reportado,42807,8632,11.37
PhD,1652,373,0.44


In [5]:
ocupaciones = I.top_valores(empleos["occupation_label"], 10)
print(f"Ocupaciones distintas con etiqueta: {int(empleos['occupation_label'].notna().sum()):,} "
      f"(de {empleos['occupation_label'].nunique():,} únicas)")
display(ocupaciones)


Ocupaciones distintas con etiqueta: 348,263 (de 2,855 únicas)


,n,%_de_con_etiqueta
occupation_label,,
administrative assistant,16864,4.84
sales assistant,10017,2.88
shop assistant,9676,2.78
warehouse worker,7106,2.04
cashier,5990,1.72
kitchen assistant,5747,1.65
waiter/waitress,5347,1.54
building cleaner,4785,1.37
factory hand,4654,1.34


In [6]:
grupos = I.top_valores(empleos["isco_group_label"], 10)
print(f"Grupos ISCO distintos con etiqueta: {empleos['isco_group_label'].nunique():,}")
display(grupos)


Grupos ISCO distintos con etiqueta: 424

,n,%_de_con_etiqueta
isco_group_label,,
Shop sales assistants,30241,8.62
Administrative and executive secretaries,24440,6.97
Freight handlers,13736,3.92
Secondary education teachers,8743,2.49
Waiters,7778,2.22
Social work associate professionals,7559,2.16
Child care workers,6590,1.88
"Cleaners and helpers in offices, hotels and other establishments",6573,1.87
Cashiers and ticket clerks,6057,1.73


In [7]:
banderas = I.resumen_banderas(empleos)
display(banderas)
sin_area_por_nivel = (
    empleos.loc[empleos["isco_group"].isna(), "university_level"].value_counts()
    / empleos["university_level"].value_counts() * 100
).round(2).sort_values(ascending=False)
print("Sin área ISCO por nivel educativo (% del nivel):")
display(sin_area_por_nivel)


,flag,n,%_del_total
0,es_unknown_ocupacion,25805,6.85
1,es_rescatado,2499,0.66
2,es_vigente,18956,5.03


Sin área ISCO por nivel educativo (% del nivel):


university_level
Bachelor            7.36
Master              7.06
No reportado        6.65
Secondary school    6.40
PhD                 5.93
Name: count, dtype: float64

**Qué deja claro el bloque 1:** el dataset cubre algo más de cinco décadas (inicios
Q1 1965–Q4 2020, fines Q1 1970–Q4 2024), en su mayoría experiencia breve (5 trimestres de
mediana), con 5 niveles educativos y 424 grupos ocupacionales (ISCO). Las banderas separan lo
que sí se sabe de lo que no: `unknown` (6,85 %), `rescatado` (0,66 %) y `vigente` (5,03 %,
final censurado).

Estas cifras están escritas a mano en este párrafo, pero **no pueden quedar desfasadas**: la
celda de línea base de arriba lleva `assert` contra la auditoría y el libro deja de ejecutarse
si deja de cuadrar. Los porcentajes de los bloques 4 y 5 sí se calculan en vivo.

## 🔎 2 · Exploración dirigida

Cada sección plantea una **pregunta** y la responde con un **filtro reutilizable**
de `src/filtro/filtros.py` o un **indicador** de `src/filtro/indicadores.py`.
Ninguna selección imputa valores: lo no clasificado se excluye y se reporta.


In [8]:
# P1 · ¿Cuántas personas tienen educación de doctorado y cuántas experiencias?
phd = F.filtrar_por_nivel_educativo(empleos, "PhD")
display(F.resumen(phd, "university_level='PhD'"))


,segmento,filas,personas,vigentes,sin_area_isco,sin_ocupacion
0,university_level='PhD',1652,373,94,98,175


In [9]:
# P2 · En el grupo ISCO más frecuente, ¿cuántas experiencias/personas hay?
grupo_top = empleos["isco_group_label"].dropna().value_counts().index[0]
sub = F.filtrar_por_grupo_isco(empleos, grupo_top)
print(f"Grupo ISCO más frecuente: {grupo_top}")
display(F.resumen(sub, f"isco_group_label='{grupo_top}'"))


Grupo ISCO más frecuente: Shop sales assistants


,segmento,filas,personas,vigentes,sin_area_isco,sin_ocupacion
0,isco_group_label='Shop sales assistants',30241,18676,1600,0,0


In [10]:
# P3 · Experiencias iniciadas 2015–2019: ¿cuántas y cuál es su composición?
recientes = F.filtrar_por_periodo(empleos, 2015, 2019)
print(f"Experiencias iniciadas 2015-2019: {len(recientes):,} "
      f"(de {personas:,} personas, {recientes['resume_id'].nunique():,} aquí)")
vigentes_corte = F.contar_banderas(recientes, "es_vigente")
print(f"- Vigentes dentro de ese corte: {vigentes_corte:,} "
      f"({100 * vigentes_corte / len(recientes):.1f} % del corte)")
display(F.resumen(recientes, "inicio 2015-2019"))


Experiencias iniciadas 2015-2019: 73,939 (de 71061 personas, 31,067 aquí)
- Vigentes dentro de ese corte: 6,312 (8.5 % del corte)


,segmento,filas,personas,vigentes,sin_area_isco,sin_ocupacion
0,inicio 2015-2019,73939,31067,6312,5252,5835


In [11]:
# P4 · ¿Vigentes vs finalizadas? (la duración solo existe para las finalizadas)
vigentes = F.filtrar_vigentes(empleos, True)
finalizadas = F.filtrar_vigentes(empleos, False)
print(f"Vigentes  : {len(vigentes):,} ({100 * len(vigentes) / filas:.1f} %)")
print(f"Finalizadas: {len(finalizadas):,} ({100 * len(finalizadas) / filas:.1f} %)")
print(f"Mediana de duración (finalizadas): {I.duracion_trimestres(finalizadas).median():.0f} trimestres")
print("Nota: las vigentes no tienen fin observado; su duración es censura, no 0.")


Vigentes  : 18,956 (5.0 %)
Finalizadas: 357,611 (95.0 %)


Mediana de duración (finalizadas): 5 trimestres
Nota: las vigentes no tienen fin observado; su duración es censura, no 0.


In [12]:
# P5 · ¿Cuánto pesa 'sin clasificar' y cómo se reparte por nivel?
sin_area = F.filtrar_sin_area_isco(empleos)
print(f"Experiencias sin área ISCO: {len(sin_area):,} "
      f"({100 * len(sin_area) / filas:.2f} %) · {sin_area['resume_id'].nunique():,} personas")
print("% de experiencias sin área dentro de cada nivel educativo:")
niv_total = empleos["university_level"].value_counts()
niv_sin = sin_area["university_level"].value_counts()
display((niv_sin / niv_total * 100).round(2).rename("%_sin_area").to_frame())


Experiencias sin área ISCO: 25,805 (6.85 %) · 16,055 personas
% de experiencias sin área dentro de cada nivel educativo:


,%_sin_area
university_level,
Bachelor,7.36
Master,7.06
No reportado,6.65
PhD,5.93
Secondary school,6.40


**Qué deja claro el bloque 2:** los filtros responden preguntas concretas y dejan ver el peso
de lo no clasificado: por ejemplo, el grupo ISCO más frecuente concentra decenas de miles de
experiencias, y el porcentaje de experiencias sin área varía entre niveles. Estos números se
usan tal cual en los bloques siguientes.


## 🧩 3 · Relaciones entre variables

Buscamos conexiones entre **nivel educativo**, **grupo ocupacional (ISCO)**, **duración** y
**trayectoria** (transiciones y periodos sin empleo registrado).

> ⚠️ **Pluriempleo y repeticiones:** el parquet limpio conserva experiencias simultáneas
> (misma persona, periodos que se cruzan) y repeticiones exactas. Para transiciones y
> solapamiento las repeticiones exactas se quitan; para brechas, el máximo de fines previos
> absorbe el traslape. Ningún conteo lo inventa el análisis: la definición se declara.


In [13]:
# Nivel educativo ↔ grupo ISCO (frecuencia relativa por fila-nivel)
tabla_nivel_grupo = I.crosstab_nivel_grupo(empleos, top=30)
display(tabla_nivel_grupo)
# Grupo más representado por nivel
print("Grupos ISCO más representados dentro de cada nivel (top 3, % del nivel):")
for nivel in ["Secondary school", "Bachelor", "Master", "PhD", "No reportado"]:
    vc = empleos.loc[empleos["university_level"].eq(nivel), "isco_group_label"] \
        .dropna().value_counts(normalize=True).head(3) * 100
    print(f"- {nivel}: " + ", ".join(f"{g} {p:.1f}%" for g, p in vc.items()))


isco_group_label,Accounting and bookkeeping clerks,Accounting associate professionals,Administrative and executive secretaries,Advertising and marketing professionals,Bartenders,Cashiers and ticket clerks,Child care workers,"Cleaners and helpers in offices, hotels and other establishments",Food and related products machine operators,Food service counter attendants,...,Retail and wholesale trade managers,Secondary education teachers,Secretaries (general),Shop sales assistants,Social work and counselling professionals,Social work associate professionals,"Supply, distribution and related managers",Teachers’ aides,University and higher education teachers,Waiters
university_level,,,,,,,,,,,,,,,,,,,,,
Bachelor,2.42,3.25,15.07,3.12,1.28,3.10,4.45,1.72,0.69,1.07,...,3.00,7.77,1.80,14.59,2.80,6.66,1.23,3.91,1.42,4.24
Master,1.92,1.80,15.10,6.93,1.23,2.41,2.42,1.18,0.61,0.86,...,3.14,14.71,1.25,11.59,4.17,5.04,1.86,2.09,4.66,3.71
No reportado,1.70,1.28,11.62,1.35,1.48,3.16,3.25,6.25,1.91,1.48,...,2.80,1.01,1.46,18.58,0.57,2.44,1.39,0.75,0.41,4.50
PhD,0.36,0.00,8.57,3.21,0.54,1.07,0.71,0.71,1.43,0.54,...,1.96,17.32,0.71,3.39,2.32,3.39,1.43,1.07,44.11,1.07
Secondary school,1.75,1.52,12.77,0.73,1.48,4.02,3.59,5.41,1.85,1.49,...,2.23,0.55,1.62,20.23,0.53,2.47,1.09,0.63,0.24,4.66


Grupos ISCO más representados dentro de cada nivel (top 3, % del nivel):


- Secondary school: Shop sales assistants 10.8%, Administrative and executive secretaries 6.8%, Freight handlers 5.7%
- Bachelor: Administrative and executive secretaries 7.8%, Shop sales assistants 7.5%, Secondary education teachers 4.0%
- Master: Administrative and executive secretaries 6.5%, Secondary education teachers 6.3%, Shop sales assistants 5.0%
- PhD: University and higher education teachers 15.9%, Biologists, botanists, zoologists and related professionals 6.6%, Secondary education teachers 6.2%
- No reportado: Shop sales assistants 9.3%, Administrative and executive secretaries 5.8%, Freight handlers 5.3%


In [14]:
# Grupo ISCO ↔ duración (mediana de trimestres por grupo, top en frecuencia)
duracion_por_grupo = I.duracion_mediana_por_grupo(empleos, top=15)
display(duracion_por_grupo)
print("Grupo con mayor mediana de duración entre los top:",
      duracion_por_grupo["mediana"].idxmax(), "·", duracion_por_grupo["mediana"].max(), "trimestres")


,mediana,n_empleos
isco_group_label,,
Shop sales assistants,5.0,28641
Administrative and executive secretaries,5.0,23091
Freight handlers,5.0,12887
Secondary education teachers,4.0,8324
Waiters,5.0,7404
Social work associate professionals,5.0,7189
Child care workers,5.0,6218
"Cleaners and helpers in offices, hotels and other establishments",5.0,6151
Cashiers and ticket clerks,5.0,5709


Grupo con mayor mediana de duración entre los top: Retail and wholesale trade managers · 9.0 trimestres


In [15]:
# Nivel educativo ↔ duración
display(I.duracion_mediana_por_nivel(empleos))


,mediana,n_empleos
university_level,,
Bachelor,5.0,123269
Master,5.0,51753
No reportado,5.0,40603
PhD,8.0,1558
Secondary school,5.0,140428


In [16]:
# Transiciones consecutivas: ¿la gente cambia de grupo ISCO entre experiencias?
trans = I.transiciones_consecutivas(empleos)
mismo = I.proporcion_mismo_grupo(empleos)
print(f"Parejas consecutivas con ambas etiquetas: {int(trans['n'].sum()):,}")
print(f"% que conservan el grupo ISCO: {mismo:.1f} %")
print("Transiciones más frecuentes (desde -> hacia):")
display(trans.head(10))


Parejas consecutivas con ambas etiquetas: 252,496
% que conservan el grupo ISCO: 18.3 %
Transiciones más frecuentes (desde -> hacia):


,desde,hacia,n
35123,Shop sales assistants,Shop sales assistants,6273
641,Administrative and executive secretaries,Administrative and executive secretaries,4826
33307,Secondary education teachers,Secondary education teachers,2892
15713,Freight handlers,Freight handlers,2185
34809,Shop sales assistants,Administrative and executive secretaries,1665
35965,Social work associate professionals,Social work associate professionals,1475
940,Administrative and executive secretaries,Shop sales assistants,1471
7008,Child care workers,Child care workers,1123
7588,"Cleaners and helpers in offices, hotels and ot...","Cleaners and helpers in offices, hotels and ot...",972
387,Accounting associate professionals,Accounting associate professionals,912


In [17]:
# % que conserva el grupo ISCO, por nivel educativo
filas_nivel = []
for nivel in ["Secondary school", "Bachelor", "Master", "PhD", "No reportado"]:
    sub = empleos[empleos["university_level"].eq(nivel)]
    trans_n = I.transiciones_consecutivas(sub)
    filas_nivel.append({
        "nivel": nivel,
        "parejas": int(trans_n["n"].sum()),
        "%_mismo_grupo": round(I.proporcion_mismo_grupo(sub), 2),
    })
tabla_mismo_nivel = pd.DataFrame(filas_nivel)
display(tabla_mismo_nivel)


,nivel,parejas,%_mismo_grupo
0,Secondary school,99770,18.85
1,Bachelor,86523,18.49
2,Master,36429,15.76
3,PhD,1096,17.06
4,No reportado,28678,18.99


In [18]:
# Primera ↔ segunda experiencia (¿dónde aterriza la gente?).
prim_seg = I.primera_a_segunda(empleos)
print(f"Personas con primera y segunda experiencia etiquetadas: "
      f"{int(prim_seg['n'].sum()):,}")
print("Pares primera->segunda más frecuentes:")
display(prim_seg.head(10))


Personas con primera y segunda experiencia etiquetadas: 58,386
Pares primera->segunda más frecuentes:


,isco_group_label_primera,isco_group_label_segunda,n
15453,Shop sales assistants,Shop sales assistants,1656
314,Administrative and executive secretaries,Administrative and executive secretaries,1216
15185,Shop sales assistants,Administrative and executive secretaries,468
7008,Freight handlers,Freight handlers,428
548,Administrative and executive secretaries,Shop sales assistants,360
14579,Secondary education teachers,Secondary education teachers,307
15874,Social work associate professionals,Social work associate professionals,290
18456,Waiters,Shop sales assistants,245
18496,Waiters,Waiters,245
2902,Cashiers and ticket clerks,Shop sales assistants,219


In [19]:
# Brechas sin empleo registrado (proxy de desempleo / inactividad)
tabla_brechas, resumen_brechas = I.brechas_entre_empleos(empleos)
print("Resumen de brechas (trimestres sin empleo registrado entre experiencias):")
display(pd.DataFrame([resumen_brechas]))
print(f"{resumen_brechas['personas_con_brecha']:,} de {resumen_brechas['personas']:,} personas "
      f"({100 * resumen_brechas['personas_con_brecha'] / resumen_brechas['personas']:.1f} %) "
      f"tienen al menos una brecha; mediana de {resumen_brechas['mediana_trimestres']:.0f} "
      f"trimestres y máximo de {resumen_brechas['max_trimestres']:.0f}.")


Resumen de brechas (trimestres sin empleo registrado entre experiencias):


,personas,personas_con_brecha,brechas,mediana_trimestres,max_trimestres
0,71061,42612,79442,3.0,155.0


42,612 de 71,061 personas (60.0 %) tienen al menos una brecha; mediana de 3 trimestres y máximo de 155.


In [20]:
# Brechas por nivel educativo (¿el desempleo es menor con más estudios?)
display(I.brechas_por_nivel(empleos))


,personas_con_brecha,mediana_brecha,personas_nivel,%_con_brecha
university_level,,,,
Bachelor,13208,3.0,22854,57.79
Master,5789,4.0,10263,56.41
No reportado,5396,5.0,8632,62.51
PhD,198,4.0,373,53.08
Secondary school,18021,4.0,28939,62.27


In [21]:
# Traslapes (mismo método del diagnóstico: inicio <= fin de la fila anterior)
traslapes = I.resumen_personas_solapadas(empleos)
display(traslapes)
pct_tras = float(traslapes.loc[traslapes["metrica"].eq("%_personas"), "valor"].iloc[0])
print(f"{pct_tras:.1f} % de las personas tienen al menos un traslape: "
      "se interpreta como pluriempleo real o dato impreciso, no como brecha.")


,metrica,valor
0,filas_traslapadas,187164.00
1,personas_traslapadas,58320.00
2,personas_total,71061.00
3,%_personas,82.07


82.1 % de las personas tienen al menos un traslape: se interpreta como pluriempleo real o dato impreciso, no como brecha.


**Qué deja claro el bloque 3:** el nivel educativo se asocia con el repertorio de grupos ISCO
y con la permanencia: conservar el mismo grupo entre experiencias es minoría (menos de 1 de
cada 5 transiciones), mientras que las brechas sin empleo registrado son mayoritarias (>50 % de
las personas) y parecen menores entre quienes reportan más estudios. Ninguna cifra se escribió a
mano: todas vienen de las celdas de código de este bloque.


## 🧠 4 · Mini investigación

Cuatro preguntas guiadas, cada una con el formato **Pregunta → Filtro/datos → Resultado →
Interpretación**. La última retoma el insight clave del análisis de brechas: el historial no
solo dice *en qué* trabajó cada persona, también *en qué momentos* no había empleo registrado
(proxy de estar o no desempleado en ese lapso).

### Q1 · ¿La concentración ocupacional cambia con el nivel educativo?


In [22]:
# Pregunta: ¿cambia la concentración de grupos ISCO según el nivel educativo?
# Dato: top 3 grupos por nivel (% de experiencias con etiqueta dentro del nivel).
for nivel in ["Secondary school", "Bachelor", "Master", "PhD", "No reportado"]:
    vc = empleos.loc[empleos["university_level"].eq(nivel), "isco_group_label"] \
        .dropna().value_counts(normalize=True).head(3) * 100
    print(f"{nivel}: " + ", ".join(f"{g} = {p:.1f} %" for g, p in vc.items()))
print()
print("INTERPRETACIÓN: observar si el top de cada nivel es el mismo o cambia el repertorio "
      "ocupacional; un top distinto por nivel sugiere que la educación ordena el destino laboral.")


Secondary school: Shop sales assistants = 10.8 %, Administrative and executive secretaries = 6.8 %, Freight handlers = 5.7 %
Bachelor: Administrative and executive secretaries = 7.8 %, Shop sales assistants = 7.5 %, Secondary education teachers = 4.0 %
Master: Administrative and executive secretaries = 6.5 %, Secondary education teachers = 6.3 %, Shop sales assistants = 5.0 %
PhD: University and higher education teachers = 15.9 %, Biologists, botanists, zoologists and related professionals = 6.6 %, Secondary education teachers = 6.2 %
No reportado: Shop sales assistants = 9.3 %, Administrative and executive secretaries = 5.8 %, Freight handlers = 5.3 %

INTERPRETACIÓN: observar si el top de cada nivel es el mismo o cambia el repertorio ocupacional; un top distinto por nivel sugiere que la educación ordena el destino laboral.


### Q2 · ¿Cuánto dura una experiencia y cambia por nivel educativo?


In [23]:
# Dato: distribución global y por nivel.
display(I.distribucion_duracion(empleos))
display(I.duracion_mediana_por_nivel(empleos))
print("INTERPRETACIÓN: mediana e IQR muestran la duración típica; la comparación por nivel "
      "cuantifica si estudiar más prolonga la permanencia en cada empleo.")


,n_finalizadas,mediana,media,q1,q3,iqr,min,max,fuera_de_iqr_n,fuera_de_iqr_%
0,357611,5.0,9.96,2.0,11.0,9.0,2.0,160.0,35477,9.92


,mediana,n_empleos
university_level,,
Bachelor,5.0,123269
Master,5.0,51753
No reportado,5.0,40603
PhD,8.0,1558
Secondary school,5.0,140428


INTERPRETACIÓN: mediana e IQR muestran la duración típica; la comparación por nivel cuantifica si estudiar más prolonga la permanencia en cada empleo.


### Q3 · ¿Qué tan lineal es la carrera (mismo grupo ISCO entre experiencias)?


In [24]:
display(tabla_mismo_nivel)
mismo_global = I.proporcion_mismo_grupo(empleos)
print(f"INTERPRETACIÓN: a nivel global solo {mismo_global:.1f} % de las transiciones conservan "
      "el grupo ISCO; por nivel se ve si ese apego crece o cae. Una carrera 'lineal' sería la "
      "excepción, no la regla.")


,nivel,parejas,%_mismo_grupo
0,Secondary school,99770,18.85
1,Bachelor,86523,18.49
2,Master,36429,15.76
3,PhD,1096,17.06
4,No reportado,28678,18.99


INTERPRETACIÓN: a nivel global solo 18.3 % de las transiciones conservan el grupo ISCO; por nivel se ve si ese apego crece o cae. Una carrera 'lineal' sería la excepción, no la regla.


### Q4 · ¿Cuántos periodos sin empleo registrado hay y cómo cambian por nivel educativo?


In [25]:
# Dato: brechas por persona y por nivel.
display(pd.DataFrame([resumen_brechas]))
brechas_nivel = I.brechas_por_nivel(empleos)
display(brechas_nivel)
print("INTERPRETACIÓN: la proporción de personas con al menos una brecha y su mediana "
      "aproximan cuánto tiempo pasa alguien sin empleo registrado entre experiencias. "
      "La comparación por nivel sugiere si la educación se asocia con periodos de "
      "desocupación más cortos.")


,personas,personas_con_brecha,brechas,mediana_trimestres,max_trimestres
0,71061,42612,79442,3.0,155.0


,personas_con_brecha,mediana_brecha,personas_nivel,%_con_brecha
university_level,,,,
Bachelor,13208,3.0,22854,57.79
Master,5789,4.0,10263,56.41
No reportado,5396,5.0,8632,62.51
PhD,198,4.0,373,53.08
Secondary school,18021,4.0,28939,62.27


INTERPRETACIÓN: la proporción de personas con al menos una brecha y su mediana aproximan cuánto tiempo pasa alguien sin empleo registrado entre experiencias. La comparación por nivel sugiere si la educación se asocia con periodos de desocupación más cortos.


**Cierre de la mini investigación:** cada pregunta produjo un número con su lectura; las
interpretaciones conectan el dato con la hipótesis de negocio (reclutar/estudiar/la empleabilidad
según nivel). Los resultados exactos quedan en las celdas anteriores.


## 🧭 5 · Hallazgos (con DATO OBSERVADO · INTERPRETACIÓN · LIMITACIÓN)

Cada hallazgo calcula su dato **en vivo** — no hay cifras copiadas.


In [26]:
# Hallazgo 1 · Concentración ocupacional
top3 = I.top_valores(empleos["occupation_label"], 3)
top3_share = float(top3["%_de_con_etiqueta"].sum())
top1_ocu = str(top3.index[0])
display(top3)
print(f"DATO OBSERVADO: las 3 ocupaciones más frecuentes concentran {top3_share:.1f} % "
      "de las experiencias con etiqueta.")
print(f"INTERPRETACIÓN: el mercado laboral de estos CV se apoya en pocos roles de servicio "
      f"y administrativos (la más común es '{top1_ocu}').")
print(f"LIMITACIÓN: la etiqueta depende del emparejamiento de la fuente; "
      f"{100 * len(sin_area) / len(empleos):.2f} % no tiene área ISCO y no entra en este "
      f"total ({len(sin_area):,} filas, se reporta por separado).")


,n,%_de_con_etiqueta
occupation_label,,
administrative assistant,16864,4.84
sales assistant,10017,2.88
shop assistant,9676,2.78


DATO OBSERVADO: las 3 ocupaciones más frecuentes concentran 10.5 % de las experiencias con etiqueta.
INTERPRETACIÓN: el mercado laboral de estos CV se apoya en pocos roles de servicio y administrativos (la más común es 'administrative assistant').
LIMITACIÓN: la etiqueta depende del emparejamiento de la fuente; 6.85 % no tiene área ISCO y no entra en este total (25,805 filas, se reporta por separado).


In [27]:
# Hallazgo 2 · Duración típica: experiencias cortas, con cola larga
d = I.distribucion_duracion(empleos)
display(d)
print(f"DATO OBSERVADO: mediana de {d.loc[0, 'mediana']:.0f} trimestres, IQR {d.loc[0, 'q1']:.0f}–"
      f"{d.loc[0, 'q3']:.0f}; {d.loc[0, 'fuera_de_iqr_%']:.0f} % fuera del IQR (Tukey).")
print("INTERPRETACIÓN: la rotación es alta; la cola larga (experiencias que superan por "
      f"mucho la mediana, hasta {d.loc[0, 'max']:.0f} trimestres) son carreras de larga data.")
print(f"LIMITACIÓN: IQR y mediana se calculan sobre empleos finalizados; las "
      f"{vigentes_present:,} vigentes son censura y no entran.")


,n_finalizadas,mediana,media,q1,q3,iqr,min,max,fuera_de_iqr_n,fuera_de_iqr_%
0,357611,5.0,9.96,2.0,11.0,9.0,2.0,160.0,35477,9.92


DATO OBSERVADO: mediana de 5 trimestres, IQR 2–11; 10 % fuera del IQR (Tukey).
INTERPRETACIÓN: la rotación es alta; la cola larga (experiencias que superan por mucho la mediana, hasta 160 trimestres) son carreras de larga data.
LIMITACIÓN: IQR y mediana se calculan sobre empleos finalizados; las 18,956 vigentes son censura y no entran.


In [28]:
# Hallazgo 3 · La mayoría cambia de grupo ocupacional entre experiencias
m = I.proporcion_mismo_grupo(empleos)
print(f"DATO OBSERVADO: solo {m:.1f} % de las transiciones consecutivas conservan el grupo ISCO.")
print("INTERPRETACIÓN: las trayectorias laborales son poco lineales; cambiar de área es lo común.")
print("LIMITACIÓN: la transición solo se observa entre experiencias etiquetadas (se excluyen "
      "las unknown y las vigentes sin continuidad); el orden usa inicio/fin.")


DATO OBSERVADO: solo 18.3 % de las transiciones consecutivas conservan el grupo ISCO.
INTERPRETACIÓN: las trayectorias laborales son poco lineales; cambiar de área es lo común.
LIMITACIÓN: la transición solo se observa entre experiencias etiquetadas (se excluyen las unknown y las vigentes sin continuidad); el orden usa inicio/fin.


In [29]:
# Hallazgo 4 · Periodos sin empleo registrado (desempleo proxy) son la norma
r = resumen_brechas
print(f"DATO OBSERVADO: {r['personas_con_brecha']:,} de {r['personas']:,} personas "
      f"({100 * r['personas_con_brecha'] / r['personas']:.1f} %) tienen ≥1 brecha; "
      f"mediana {r['mediana_trimestres']:.0f} trimestres.")
print("INTERPRETACIÓN: entre experiencias hay lapsos frecuentes sin empleo registrado; el "
      "dato responde a la pregunta de si la persona 'estuvo o no desempleada' en ese tiempo.")
print("LIMITACIÓN: 'sin empleo registrado' ≠ desempleo confirmado: puede ser estudio, "
      "inactividad, trabajo no declarado o vacío del currículum; además 5 % de las filas están "
      "censuradas (Present). Es un proxy, no una medida de desempleo.")


DATO OBSERVADO: 42,612 de 71,061 personas (60.0 %) tienen ≥1 brecha; mediana 3 trimestres.
INTERPRETACIÓN: entre experiencias hay lapsos frecuentes sin empleo registrado; el dato responde a la pregunta de si la persona 'estuvo o no desempleada' en ese tiempo.
LIMITACIÓN: 'sin empleo registrado' ≠ desempleo confirmado: puede ser estudio, inactividad, trabajo no declarado o vacío del currículum; además 5 % de las filas están censuradas (Present). Es un proxy, no una medida de desempleo.


In [30]:
# Hallazgo 5 · Las fronteras del dato: unknown y censura a mano derecha
display(I.resumen_banderas(empleos))
print(f"DATO OBSERVADO: {100 * len(sin_area) / len(empleos):.2f} % sin área ISCO y "
      f"{100 * vigentes_present / len(empleos):.2f} % de experiencias vigentes (fin = Present).")
print("INTERPRETACIÓN: el análisis distintivo depende de que no se impute nada de esto; "
      "la interpretación vive en declarar las exclusiones.")
print("LIMITACIÓN: cualquier conclusión sobre ocupación aplica al 93 % etiquetado, y sobre "
      "duración al 95 % con fin observado.")


,flag,n,%_del_total
0,es_unknown_ocupacion,25805,6.85
1,es_rescatado,2499,0.66
2,es_vigente,18956,5.03


DATO OBSERVADO: 6.85 % sin área ISCO y 5.03 % de experiencias vigentes (fin = Present).
INTERPRETACIÓN: el análisis distintivo depende de que no se impute nada de esto; la interpretación vive en declarar las exclusiones.
LIMITACIÓN: cualquier conclusión sobre ocupación aplica al 93 % etiquetado, y sobre duración al 95 % con fin observado.


## 🧩 6 · Perfil de competencias (atribución por contrato)

Hasta aquí el libro contó **cuántas** competencias exigía cada cargo. Este bloque les pone
**nombre** y las ata a una persona concreta.

Las competencias de ESCO son un atributo del **oficio**, no del individuo: describen lo que
un cargo **exige**, no lo que alguien **sabe**. Por eso el nombre de una competencia es un
hecho del puesto, y por eso la lectura correcta es *«ocupó ese cargo → su contrato obligaba
a esas competencias»* (**atribución por contrato**). No es una medición de las capacidades de
la persona: el dato no las observa, y presentarlas como si lo hiciera sería inventar.

Para que eso se vea sin inflar el dataset, los nombres viven en una **tabla aparte**, con una
fila por (oficio, competencia), y se unen por `occupation_code` con un cruce N:1.

In [31]:
# La tabla de competencias: 1 fila = 1 competencia de 1 oficio.
competencias = I.cargar_competencias()

print(f"competencias_por_oficio.csv: {len(competencias):,} filas x {competencias.shape[1]} columnas")
print(f"  {competencias['occupation_code'].nunique():,} oficios | {competencias['habilidad_nombre'].nunique():,} habilidades distintas")
print(f"  tipos: {competencias['habilidad_tipo'].value_counts().to_dict()}")

# El cruce con el dataset es N:1 y no multiplica nada: una fila del
# dataset recibe N competencias, una fila de la tabla recibe 1.
print(f"\nCruce N:1 por occupation_code: las {len(empleos):,} experiencias siguen siendo {len(empleos):,}.")
print(f"La tabla aporta {len(competencias):,} nombres para {competencias['occupation_code'].nunique():,} oficios.")

display(competencias.head(6))

competencias_por_oficio.csv: 67,600 filas x 4 columnas
  3,039 oficios | 11,378 habilidades distintas
  tipos: {'skill/competence': 51155, 'knowledge': 16406, 'sin_tipo': 39}

Cruce N:1 por occupation_code: las 376,567 experiencias siguen siendo 376,567.
La tabla aporta 67,600 nombres para 3,039 oficios.


,occupation_code,occupation_label,habilidad_nombre,habilidad_tipo
0,0110.1,air force officer,apply air force procedures,skill/competence
1,0110.1,air force officer,apply military aviation regulations,skill/competence
2,0110.1,air force officer,collect real-time geolocation information,skill/competence
3,0110.1,air force officer,devise military tactics,skill/competence
4,0110.1,air force officer,ensure compliance with types of weapons,skill/competence
5,0110.1,air force officer,ensure public safety and security,skill/competence


In [32]:
# La regla de LECTURA: se agrupa por oficio y el cargo trae todas sus
# competencias. "1 fila = 1 competencia" era la regla de GUARDADO.
por_oficio = I.competencias_por_oficio(competencias)

oficio = ("domestic cleaner", "9111.1")
competencias_del_cargo = por_oficio[oficio[1]]
print(f"{oficio[0]} ({oficio[1]}) exige {len(competencias_del_cargo)} competencias:")
for nombre in competencias_del_cargo[:8]:
    print(f"  · {nombre}")
print(f"  · ... (+{len(competencias_del_cargo) - 8} más)")

domestic cleaner (9111.1) exige 20 competencias:
  · clean glass surfaces
  · clean household linens
  · clean rooms
  · clean surfaces
  · clean toilet facilities
  · cleaning techniques
  · ensure cleanliness of food preparation area
  · handle chemical cleaning agents
  · ... (+12 más)


In [33]:
# El perfil de una persona: educación + ruta + lo que exigía cada contrato.
persona = "100013"
display(I.resumen_perfil(empleos, persona))
perfil = I.perfil_persona(empleos, competencias, persona)
for f in perfil.itertuples():
    veces = f"x{int(f.veces_ese_oficio):g}" if pd.notna(f.veces_ese_oficio) else ""
    print(f"{f.orden:>2}. {f.occupation_label:<34} {f.start_date} -> {f.end_date:<9} "
          f"{veces:<4} {f.n_competencias} competencias")
    if f.aviso:
        print(f"      {f.aviso}")
        continue
    for nombre in f.competencias[:4]:
        print(f"      · {nombre}")
    if f.n_competencias > 4:
        print(f"      · ... (+{f.n_competencias - 4} más)")

,persona,nivel_educativo,periodos,oficios_distintos,periodos_con_oficio,periodos_sin_oficio,contrato_mas_repetido
0,100013,No reportado,3,3,3,0,1


 1. domestic cleaner                   Q1 2007 -> Q1 2008   x1   20 competencias
      · clean glass surfaces
      · clean household linens
      · clean rooms
      · clean surfaces
      · ... (+16 más)
 2. stop-motion animator               Q1 2007 -> Q1 2010   x1   10 competencias
      · adapt to type of media
      · analyse a script
      · develop animations
      · finish project within budget
      · ... (+6 más)
 3. building cleaner                   Q1 2009 -> Q1 2012   x1   24 competencias
      · arrange furniture
      · clean building facade
      · clean building floors
      · clean glass surfaces
      · ... (+20 más)


In [34]:
# Cuando el cargo no se pudo clasificar no se inventa nada: se declara
# DESCONOCIDO, conservando las fechas del periodo. Poner cero competencias
# sería una afirmación falsa sobre el 7,5 % del dataset sin oficio.
persona = "299581"
display(I.resumen_perfil(empleos, persona))
perfil = I.perfil_persona(empleos, competencias, persona)
sin_clasificar = perfil.loc[perfil["aviso"].ne("")]

print(f"{perfil['occupation_code'].notna().sum()} periodos con oficio clasificado | "
      f"{len(sin_clasificar)} sin clasificar (no son cero competencias: son desconocidas)")

display(sin_clasificar[["orden", "occupation_label", "start_date", "end_date",
                        "saber_skills", "aviso"]])

,persona,nivel_educativo,periodos,oficios_distintos,periodos_con_oficio,periodos_sin_oficio,contrato_mas_repetido
0,299581,Master,20,16,17,3,2


17 periodos con oficio clasificado | 3 sin clasificar (no son cero competencias: son desconocidas)


,orden,occupation_label,start_date,end_date,saber_skills,aviso
6,7,[OFICIO NO CLASIFICADO],Q1 2007,Q2 2007,no_clasificado,⚠ habilidades desconocidas: no sabemos que tra...
15,16,[OFICIO NO CLASIFICADO],Q1 2010,Q3 2012,no_clasificado,⚠ habilidades desconocidas: no sabemos que tra...
16,17,[OFICIO NO CLASIFICADO],Q3 2010,Q2 2011,no_clasificado,⚠ habilidades desconocidas: no sabemos que tra...


In [35]:
# ¿Cuánto cambian los requisitos entre dos contratos consecutivos?
# Jaccard = |A ∩ B| / |A ∪ B| sobre los NOMBRES de las competencias.
detalle, resumen = I.salto_competencias(empleos, competencias)

# La serie sale mixta (enteros y flotantes) y `to_frame` la castea toda a
# float, lo que imprimiría 63748.00000. Se formatea fila por fila.
display(pd.DataFrame(
    [(clave, f"{valor:,.2f}" if isinstance(valor, float) else f"{valor:,}")
     for clave, valor in resumen.items()],
    columns=["indicador", "valor"],
))

,indicador,valor
0,personas,"63,748"
1,transiciones,"249,323"
2,mismo_cargo,"33,772"
3,%_mismo_cargo,13.55
4,jaccard_mediana,0.00
5,jaccard_p25,0.00
6,jaccard_p75,0.09
7,transiciones_sin_ninguna_compartida,"133,285"
8,%_sin_ninguna_compartida,53.46


In [36]:
# Los dos extremos del mismo indicador.
# El 13,6 % de las transiciones se queda en el mismo cargo: el top-N sin
# filtrar devolvería siempre la misma fila repetida. Se mira el cambio de oficio.
cambios = detalle.loc[detalle["desde_code"].ne(detalle["hacia_code"])]

print(f"{len(cambios):,} cambios de oficio (de {len(detalle):,} transiciones; "
      f"los {len(detalle) - len(cambios):,} restantes son el mismo cargo).")

print("\nLos 5 cambios con MÁS competencias compartidas:")
display(cambios.nlargest(5, "n_compartidas")[
    ["desde_label", "hacia_label", "n_desde", "n_hacia", "n_compartidas", "jaccard"]
].round(3).reset_index(drop=True))

print("Y 5 cambios que no comparten NINGUNA competencia:")
display(detalle.loc[detalle["n_compartidas"].eq(0),
                   ["desde_label", "hacia_label", "n_desde", "n_hacia", "jaccard"]]
        .head(5).reset_index(drop=True))

215,551 cambios de oficio (de 249,323 transiciones; los 33,772 restantes son el mismo cargo).

Los 5 cambios con MÁS competencias compartidas:


,desde_label,hacia_label,n_desde,n_hacia,n_compartidas,jaccard
0,child day care worker,child welfare worker,73,70,68,0.907
1,mental health social worker,family social worker,76,71,67,0.838
2,youth offending team worker,youth worker,72,72,67,0.870
3,residential childcare worker,child day care worker,67,73,66,0.892
4,child day care worker,residential childcare worker,73,67,66,0.892


Y 5 cambios que no comparten NINGUNA competencia:


,desde_label,hacia_label,n_desde,n_hacia,jaccard
0,public relations manager,communications lecturer,27,32,0.0
1,communications lecturer,promotion assistant,32,8,0.0
2,promotion assistant,broadcast news editor,8,19,0.0
3,domestic cleaner,stop-motion animator,20,10,0.0
4,stop-motion animator,building cleaner,10,24,0.0


**Qué deja claro el bloque 6:** con la tabla de competencias, un cargo deja de ser una etiqueta
y pasa a ser **una lista de requisitos legible**.

- **DATO OBSERVADO:** 67.600 competencias para 3.039 oficios y 11.378 habilidades distintas;
  su reparto es de cola larga (las 100 más frecuentes cubren el 17,2 % de los pares). En
  249.323 transiciones consecutivas entre contratos, la mediana del solapamiento es **0,00** y
  el **53,5 %** no comparte ni una sola competencia.
- **INTERPRETACIÓN:** las competencias son específicas de cada puesto, no un repertorio
  general. Y más de la mitad de los cambios de cargo **dejan intacto** el conjunto de
  requisitos: cambiar de oficio no es siempre cambiar de exigencia.
- **LIMITACIÓN:** esto describe **puestos, no personas**. La atribución es por contrato —el
  cargo obligaba a esas competencias—, no una medición de lo que alguien sabe. Y donde el
  oficio no se pudo clasificar, las competencias quedan **desconocidas**, nunca en cero.

## Limitaciones transversales

- **Dato auto-reportado**: los CV no son un registro oficial; la ausencia de un empleo es
  silencio, no certeza de desempleo.
- **Pluriempleo y traslapes**: la definición de "transición" depende de ordenar por inicio/fin;
  el método se declara en cada indicador y puede variar ligeramente frente a otras definiciones.
- **Sin campo de estudio**: responder "¿el empleo se relaciona con lo que estudió?" es solo
  posible por proxy (nivel educativo ↔ grupo ISCO heredado).
- **Censura**: los vigentes no tienen duración; asumimos continuidad solo hasta el corte.

## Reproducibilidad

1. Activar el entorno con `requirements.txt`.
2. Importar el notebook: *Kernel → Restart & Run All*.
3. Filtros e indicadores viven en `src/filtro/` (snake_case, funciones puras y asserts).
4. Las cifras de la línea base deben coincidir con las del `README.md`. La auditoría de la rama (`docs/AUDITORIA_RAMA_PRUEBA.md`) está **fuera de git** y documenta una versión anterior del pipeline (era `train`, 1.506.445 filas); sirve como contexto, no como fuente de verdad.
